### ISSUE TITLE
Merging VCS repository override of published package fails

### ISSUE THREAD

**Issue by bd808** [wikimedia/composer-merge-plugin#95]
First reported by @ghprod in https://github.com/wikimedia/composer-merge-plugin/issues/71#issuecomment-154435583

**composer.json:**
```
{
    "require": {
        "wikimedia/composer-merge-plugin": "dev-master"
    },
    "config": {
        "preferred-install": "dist"
    },
    "extra": {
        "merge-plugin": {
            "include": "composer.local.json"
        }
    }
}
```

**composer.local.json:**
```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/ghprod/laravel-menu"
        }
    ],
    "require":{
        "lavary/laravel-menu": "dev-master#956e50f"
    }
}
```

**Result:**
```
$ composer -vvv update
[...snip...]
Loading plugin Wikimedia\Composer\MergePlugin
  [merge-plugin] Loading composer.local.json...
  [merge-plugin] Adding vcs repository github.com/ghprod/laravel-menu
  [merge-plugin] Merging lavary/laravel-menu
Loading composer repositories with package information
[...snip...]
Downloading https://api.github.com/repos/ghprod/laravel-menu
Downloading https://api.github.com/repos/ghprod/laravel-menu/contents/composer.json?ref=master
Downloading https://api.github.com/repos/ghprod/laravel-menu/commits/master
Downloading https://api.github.com/repos/ghprod/laravel-menu/tags?per_page=100
Downloading https://api.github.com/repos/ghprod/laravel-menu/git/refs/heads?per_page=100
[...snip...]
Reading composer.json of lavary/laravel-menu (master)
Downloading https://api.github.com/repos/ghprod/laravel-menu/contents/composer.json?ref=956e50fbf0d99b2804539069b012ad52b4851dd2
Downloading https://api.github.com/repos/ghprod/laravel-menu/commits/956e50fbf0d99b2804539069b012ad52b4851dd2
[...snip...]
Downloading http://packagist.org/p/lavary/laravel-menu%24cb8432bd74b707af0300352a1d49742389810f45e5646550e5dca3a823fe4c55.json
[...snip...]
Analyzed 1426 packages to resolve dependencies
Analyzed 1200 rules to resolve dependencies
[...snip...]
  - Installing lavary/laravel-menu (dev-master bebce99)
Downloading https://api.github.com/repos/lavary/laravel-menu/zipball/bebce996ef08cf911da0139c30a2da03a2768326
```

The installed revision (bebce99) is the head of the dev branch of the project as listed on Packagist. The alternate VCS repository was not used.

If the require for "lavary/laravel-menu" is moved to composer.json, results are slightly different but still not correct. In that case the logs report:
```
  - Installing lavary/laravel-menu (dev-master 956e50f)
Downloading https://api.github.com/repos/lavary/laravel-menu/zipball/956e50f
```

This download is still from the wrong repository in this case and the files fetched from GitHub are not not actually for the desired hash.

If the "repositories" are moved to composer.json instead, the new output is:
```
  - Installing lavary/laravel-menu (dev-master 956e50f)
Downloading https://api.github.com/repos/ghprod/laravel-menu/zipball/956e50fbf0d99b2804539069b012ad52b4851dd2
```

This is the desired result for all three tests.

I think the issue here is the order that the repositories are added to `Composer\Repository\RepositoryManager`. It may take an upstream patch to fix this properly.

**Comment by ghprod** [comment]
Wow .. nice .. Thanks a lot for look at this :+1:

**Comment by bd808** [comment]
> It may take an upstream patch to fix this properly.

Fixing this is going to require working upstream with Composer to allow changing the order of repositories. The load order is well described in the [documentation](https://getcomposer.org/doc/04-schema.md#repositories) as being order dependent with the default Packagist repo being added last. Since repositories are only allowed in the root package there is currently no method provided by the internal Composer API to alter order of the list of configured repositories after they have been initially added to `RepositoryManager`'s internal private collection. We can inject new repositories into the manager but they always come after the default Packagist repo.

There may also be an additional bug related to using `#<ref>` explicit references in merged files. I need to do more testing in the absence of the current repository bug to confirm or refute that.

**Comment by ghprod** [comment]
Thanks @bd808  :+1:

**Comment by xGouley** [comment]
:+1: for this needed feature! Lost my day found that my problem is the same than yours. No other way to bypass packagist only for the wanted package?

**Comment by xGouley** [comment]
Writing my previous comment, I found a solution! In the main composer.json, you have to add: 
"repositories": [
        {
            "packagist": false
        }
],

And then in your composer.local.json, add this at the end of the repositories:
"repositories": [
        ...,
        ...,
        {
        	"type": "composer",
        	"url": "packagist.org"
        }
]

This bypass will put back packagist at the end of the list...
Could you make the SAME in the PHP code of your composer module ? In this case there is no need to have a new API call :) Maybe just an option to activate to do this automatically...

**Comment by bd808** [comment]
@xGouley wrote:
> This bypass will put back packagist at the end of the list...
Could you make the SAME in the PHP code of your composer plugin ? In this case there is no need to have a new API call :) Maybe just an option to activate to do this automatically...

The difficulty is that our plugin is not enabled until after the root composer.json has been loaded by Composer. This is intrinsic as the root package is where composer-merge-plugin is installed and configured. The addition of the default packagist repository is done as [an integral part of the root package loading process](https://github.com/composer/composer/blob/966a9827382b59064255579f524e3904527abdbe/src/Composer/Package/Loader/RootPackageLoader.php#L124-L128) so by the time composer-merge-plugin is active and able to intercept operations the repository chain has already been created. The method you discovered will work but as noted requires additional configuration of the root package by the package developer. It can certainly be used by projects that are aware of the need but a more robust solution really will require work with Composer upstream to provide a new API for interacting with the RepositoryManager.

**Comment by xGouley** [comment]
Ok I understand the problem, you have no rights to add by yourself the "packagist: false" on the root json.
So I will wait for this ticket to be solved, to fix the configuration on my side :)
TRha

**Comment by Furgas** [comment]
You could use `$repoManager->prependRepository($repo);` in `ExtraPackage::addRepositories` method instead of `$repoManager->addRepository($repo);`. You could also introduce plugin config option to control this behaviour (add or prepend).

**Comment by Furgas** [comment]
Something like [this](https://github.com/Furgas/composer-merge-plugin/commit/4a546367f5a0e15de0e37ebdbe6f8b488d262e6f). Tested and works. No PR (yet) because I don't have time now to write proper tests for it.

**Comment by bd808** [comment]
Enabling upstream change: https://github.com/composer/composer/commit/ea30392105a769c18af62891a02d3e679502010d
This commit was included in Composer 1.0.0-beta1. I think moving to requiring Composer >=1.0.0 is reasonable.

**Comment by bd808** [comment]
Let's leave this open until #124 is merged.
